test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,424 @@
+        wtxid = int(tx.getwtxid(), 16)
+        peer.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=wtxid)]))
+        self.fastforward(TXREQUEST_TIME_SKIP)
+        peer.wait_for_getdata([wtxid])
+        peer.send_and_ping(msg_tx(tx))
+
+    @cleanup
+    def test_arrival_timing_orphan(self):
+        self.log.info("Test missing parents that arrive during delay are not requested")
+        node = self.nodes[0]
+        tx_parent_arrives = self.wallet.create_self_transfer()
+        tx_parent_doesnt_arrive = self.wallet.create_self_transfer()
+        # Fake orphan spends nonexistent outputs of the two parents
+        tx_fake_orphan = self.wallet.create_self_transfer_multi(utxos_to_spend=[
+            {"txid": tx_parent_doesnt_arrive["txid"], "vout": 10, "value": tx_parent_doesnt_arrive["new_utxo"]["value"]},
+            {"txid": tx_parent_arrives["txid"], "vout": 10, "value": tx_parent_arrives["new_utxo"]["value"]}
+        ])
+
+        peer_spy = node.add_p2p_connection(PeerTxRelayer())
+        peer_normal = node.add_p2p_connection(PeerTxRelayer())
+        # This transaction is an orphan because it is missing inputs.
+        # The node should not immediately respond with a request for orphan parents.
+        # Also, no request should be sent later because it will be resolved by
+        # the time the request is scheduled to be sent.
+        peer_spy.assert_no_immediate_response(msg_tx(tx_fake_orphan["tx"]))
+
+        # Node receives transaction. It attempts to obfuscate the exact timing at which this
+        # transaction entered its mempool. Send unsolicited because otherwise we need to wait for
+        # request delays.
+        peer_normal.send_and_ping(msg_tx(tx_parent_arrives["tx"]))
+
+        # Spy peer should not be able to query the node for the parent yet, since it hasn't been
+        # announced / insufficient time has elapsed.
+        parent_inv = CInv(t=MSG_WTX, h=int(tx_parent_arrives["tx"].getwtxid(), 16))
+        assert_equal(len(peer_spy.get_invs()), 0)
+        peer_spy.assert_no_immediate_response(msg_getdata([parent_inv]))
+
+        # Request would be scheduled with this delay because it is not a preferred relay peer.
+        self.fastforward(NONPREF_PEER_TX_DELAY)
+        peer_spy.assert_never_requested(int(tx_parent_arrives["txid"], 16))
+        peer_spy.assert_never_requested(int(tx_parent_doesnt_arrive["txid"], 16))
+        # Request would be scheduled with this delay because it is by txid.
+        self.fastforward(TXID_RELAY_DELAY)
+        peer_spy.wait_for_parent_requests([int(tx_parent_doesnt_arrive["txid"], 16)])
+        peer_spy.assert_never_requested(int(tx_parent_arrives["txid"], 16))
+
+    @cleanup
+    def test_orphan_rejected_parents_exceptions(self):
+        node = self.nodes[0]
+        peer1 = node.add_p2p_connection(PeerTxRelayer())
+        peer2 = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test orphan handling when nonsegwit parent paid 0 fee")
```

Comment:
Might be better to be clearer about the different behaviours you're testing for, rather than how you're triggering those behaviours? eg:

 * Orphan handling when parent is known to be invalid
 * Orphan handling when segwit parent may be retried with alternate witness data

Reply:
Agree, updated the logging